packages/framework/immutable-merge/src/Merge.ts

```diff
@@ -1,43 +1,120 @@
 /**
- * Checks whether an argument is an object, it excludes arrays and potentially empty arrays
- * @param t the argument to check against
- * @param onlyReturnNonEmptyObjects if true only return true when the passed in arg is non-empty
+ * The basic options for recursion at a given level.  Two types for two behaviors:
+ *
+ * boolean:
+ *  - if true, recurse indefinitely
+ *  - if false, don't recurse
+ *
+ * number:
+ *  - if 0, don't recurse from this level
+ *  - if > 0, recurse that many times then stop
+ *  - if < 0, recurse indefinitely
+ */
+export type RecursionOption = boolean | number;
+
+/**
+ * a function that can be set to merge arguments
+ */
+export type CustomRecursionHandler = (...vals: any[]) => any;
+
+/**
+ * built in handler functions that can be applied for a given key
+ */
+export type BuiltinRecursionHandlers = 'appendArray';
+
+/**
+ * Handlers for recursion of a given key or type.  These can either be functions or a reference to a supported
+ * built-in merge routine
  */
-function _isNonArrayObject(t: any, onlyReturnNonEmptyObjects?: boolean): boolean {
-  return t && typeof t === 'object' && !Array.isArray(t) && (!onlyReturnNonEmptyObjects || Object.getOwnPropertyNames(t).length > 0);
+export type RecursionHandler = BuiltinRecursionHandlers | CustomRecursionHandler;
+
+/**
+ * configuration object for the merge, key names are matched with a few exceptions:
+ * - object: matches non-array object types
+ * - array: matches array types
+ * - [key: string]: matches anything by name
+ */
+export interface MergeOptions {
```

Comment:
Are we not fans of the `I` prefix on interfaces and types anymore?

Reply:
It looks like Fluent/Fabric is moving away from I prefixes and will be changing everything.